xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java

```diff
@@ -0,0 +1,111 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An {@link XdsLoadReportStore} instance holds the client side load stats for a cluster.
+ */
+@ThreadSafe
+final class XdsLoadReportStore {
+
+  private final String clusterName;
+  private final ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters;
+
+  XdsLoadReportStore(String clusterName) {
+    this(clusterName, new ConcurrentHashMap<Locality, ClientLoadCounter>());
+  }
+
+  @VisibleForTesting
+  XdsLoadReportStore(String clusterName,
+      ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters) {
+    this.clusterName = checkNotNull(clusterName, "clusterName");
+    this.localityLoadCounters = checkNotNull(localityLoadCounters, "localityLoadCounters");
+  }
+
+  /**
+   * Generates a {@link ClusterStats} containing load stats in locality granularity.
+   */
+  ClusterStats generateLoadReport(Duration interval) {
+    ClusterStats.Builder statsBuilder = ClusterStats.newBuilder().setClusterName(clusterName)
+        .setLoadReportInterval(interval);
+    for (Map.Entry<Locality, XdsClientLoadRecorder.ClientLoadCounter> entry : localityLoadCounters
+        .entrySet()) {
+      XdsClientLoadRecorder.ClientLoadSnapshot snapshot = entry.getValue().snapshot();
+      statsBuilder
+          .addUpstreamLocalityStats(UpstreamLocalityStats.newBuilder()
+              .setLocality(entry.getKey())
+              .setTotalSuccessfulRequests(snapshot.callsSucceed)
+              .setTotalErrorRequests(snapshot.callsFailed)
+              .setTotalRequestsInProgress(snapshot.callsInProgress));
+    }
+    return statsBuilder.build();
+  }
+
+  /**
+   * Discard the {@link ClientLoadCounter} for the provided locality if owned by this {@link
+   * XdsLoadReportStore} to avoid map size growing infinitely. To be called at locality updates when
+   * the provided locality is no longer considered in balancer discovery response.
+   */
+  void discardClientLoadCounter(Locality locality) {
+    localityLoadCounters.remove(locality);
+  }
+
+  /**
+   * Intercepts a in-locality PickResult with load recording {@link ClientStreamTracer.Factory}.
+   */
+  PickResult interceptPickResult(PickResult pickResult, Locality locality) {
+    if (!pickResult.getStatus().isOk()) {
+      return pickResult;
+    }
+    ClientStreamTracer.Factory originFactory = pickResult.getStreamTracerFactory();
+    if (originFactory == null) {
+      originFactory = new ClientStreamTracer.Factory() {
+        @Override
+        public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata headers) {
+          return new ClientStreamTracer() {
+          };
+        }
+      };
+    }
+
+    // TODO (chengyuanzhang): use a createClientLoadCounter method instead better?
+    XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters
+        .putIfAbsent(locality, new ClientLoadCounter());
```

Comment:
There should be an `addClientLoadCounter()` to go with `discardClientLoadCounter()`. Or better, just `addLocality()` and `removeLocality()` to give out less implementation detail. The callers of `XdsLoadReportStore` doesn't have to know about any "counter".

You don't have to have an explicit `addLocality()` per se, but I would recommend you do so, because:
1. It's easier, because then the mutation of the map can be done completely in the sync context, and there is no need for `putIfAbsent()`. 
2. It's less error-prone. If you add to the map in `interceptPickResult()`, which is on the data path (per-RPC, application thread), it may race with `removeLocality()` which is on the control path (sync context). You would need to be careful about the ordering of calling `removeLocality()` vs. updating the picker, to make sure a removed locality is never passed to `interceptPickResult()` after it's removed from the map. On the other hand, if you have `addLocality()` to add to the map, even if a removed locality is passed to `interceptPickResult()`, you can simply ignore it.

Reply:
ACK.